nexus/types/src/deployment/blueprint_diff.rs

```diff
@@ -0,0 +1,804 @@
+                            zones: added,
+                        },
+                    );
+                }
+                if !modified.is_empty() {
+                    diffs.modified.insert(
+                        sled_id,
+                        BpDiffZonesModified {
+                            generation_before: before_generation,
+                            generation_after: after_generation,
+                            zones: modified,
+                        },
+                    );
+                }
+                if !errors.is_empty() {
+                    diffs.errors.insert(
+                        sled_id,
+                        BpDiffZoneErrors {
+                            generation_before: before_generation,
+                            generation_after: after_generation,
+                            errors,
+                        },
+                    );
+                }
+            } else {
+                // No `after_zones` for this `sled_id`, so `before_zones` are removed
+                assert!(removed.is_empty());
+                for zone in before_zones.zones() {
+                    removed.push(zone);
+                }
+
+                if !removed.is_empty() {
+                    diffs.removed.insert(
+                        sled_id,
+                        BpDiffZoneDetails {
+                            generation_before: Some(before_generation),
+                            generation_after: None,
+                            zones: removed,
+                        },
+                    );
+                }
+            }
+        }
+
+        // Any sleds remaining in `after` have just been added, since we remove
+        // sleds from `after`, that were also in `before`, in the above loop.
+        for (sled_id, after_zones) in after {
+            if !after_zones.zones.is_empty() {
+                diffs.added.insert(
+                    sled_id,
+                    BpDiffZoneDetails {
+                        generation_before: None,
+                        generation_after: Some(after_zones.generation),
+                        zones: after_zones
+                            .zones
+                            .into_iter()
+                            .map(|z| z.into())
+                            .collect(),
+                    },
+                );
+            }
+        }
+
+        diffs
+    }
+
+    /// Return a [`BpSledSubtable`] for the given `sled_id`
+    ///
+    /// We collate all the data from each category to produce a single table.
+    /// The order is:
+    ///
+    /// 1. Unchanged
+    /// 2. Removed
+    /// 3. Modified
+    /// 4. Added
+    ///
+    /// The idea behind the order is to (a) group all changes together
+    /// and (b) put changes towards the bottom, so people have to scroll
+    /// back less.
+    ///
+    /// Errors are printed in a more freeform manner after the table is
+    // displayed.
+
```

Comment:
```suggestion
    /// displayed.
```

(and probably also trim the newline between the doc comment and the function?)

Reply:
fixed.